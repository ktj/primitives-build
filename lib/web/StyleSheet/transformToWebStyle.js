var prefixAll=require('inline-style-prefixer/static');
var prefixedProperties=require('./prefixedProperties');
var normalizeValue=require('./normalizeValue');
var colorWithOpacity=require('../util/colorWithOpacity');
var hasOwnProperty=Object.prototype.hasOwnProperty;

var styleShortHands={
borderColor:{
borderTopColor:true,
borderRightColor:true,
borderBottomColor:true,
borderLeftColor:true},

borderRadius:{
borderTopLeftRadius:true,
borderTopRightRadius:true,
borderBottomRightRadius:true,
borderBottomLeftRadius:true},

borderStyle:{
borderTopStyle:true,
borderRightStyle:true,
borderBottomStyle:true,
borderLeftStyle:true},

borderWidth:{
borderTopWidth:true,
borderRightWidth:true,
borderBottomWidth:true,
borderLeftWidth:true},

margin:{
marginTop:true,
marginRight:true,
marginBottom:true,
marginLeft:true},

marginHorizontal:{
marginRight:true,
marginLeft:true},

marginVertical:{
marginTop:true,
marginBottom:true},

overflow:{
overflowX:true,
overflowY:true},

padding:{
paddingTop:true,
paddingRight:true,
paddingBottom:true,
paddingLeft:true},

paddingHorizontal:{
paddingRight:true,
paddingLeft:true},

paddingVertical:{
paddingTop:true,
paddingBottom:true},

textDecorationLine:{
textDecoration:true},

writingDirection:{
direction:true}};









var sortProps=function sortProps(propsArray){return propsArray.sort(function(a,b){
var expandedA=styleShortHands[a];
var expandedB=styleShortHands[b];
if(expandedA&&expandedA[b]){
return-1;
}else if(expandedB&&expandedB[a]){
return 1;
}

return a<b?-1:a>b?1:0;
});};

var resolveFlexStyle=function resolveFlexStyle(resolvedStyle,flex){


resolvedStyle.flexGrow=flex;

resolvedStyle.flexShrink=1;

resolvedStyle.flexBasis='auto';
};

var resolveShadowProperty=function resolveShadowProperty(resolvedStyle,style){
var opacity=style.shadowOpacity!=null?style.shadowOpacity:1;
var color=colorWithOpacity(style.shadowColor,opacity);
var offset=style.shadowOffset||{width:0,height:0};
var x=normalizeValue('width',offset.width||0);
var y=normalizeValue('height',offset.height||0);
var r=style.shadowRadius||0;





var rpx=normalizeValue('shadowRadius',r*3);


resolvedStyle.boxShadow=color+' '+x+' '+y+' '+rpx;
};


var mapTransform=function mapTransform(transform){
var key=Object.keys(transform)[0];
return key+'('+normalizeValue(key,transform[key])+')';
};


var resolveTransformProperty=function resolveTransformProperty(resolvedStyle,transform){

resolvedStyle.transform=transform.map(mapTransform).join(' ');
};




var transformToWebStyle=function transformToWebStyle(style){
if(!style)return style;

var propsArray=Object.keys(style);
var sortedProps=sortProps(propsArray);
var resolvedStyle={};
var hasResolvedShadow=false;
var needsPrefix=false;

for(var i=0;i<sortedProps.length;i++){
var key=sortedProps[i];
var value=style[key];
if(!needsPrefix&&prefixedProperties[key]){
needsPrefix=true;
}
switch(key){
case'flex':
resolveFlexStyle(resolvedStyle,value,style);
break;
case'textAlignVertical':
resolvedStyle.verticalAlign=value==='center'?'middle':value;
break;
case'transform':
resolveTransformProperty(resolvedStyle,value);
break;
case'shadowColor':
case'shadowOffset':
case'shadowOpacity':
case'shadowRadius':
if(!hasResolvedShadow){
resolveShadowProperty(resolvedStyle,style);
hasResolvedShadow=true;
}
break;
case'borderColor':
case'borderRadius':
case'borderStyle':
case'borderWidth':
case'margin':
case'marginHorizontal':
case'marginVertical':
case'overflow':
case'padding':
case'paddingHorizontal':
case'paddingVertical':
case'textDecorationLine':
case'writingDirection':{
var expandedProps=styleShortHands[key];
var normalizedVal=normalizeValue(key,value);
for(var propName in expandedProps){
if(hasOwnProperty.call(expandedProps,propName)){
resolvedStyle[propName]=normalizedVal;
}
}
}break;
case'elevation':

break;
default:
resolvedStyle[key]=normalizeValue(key,value);
break;}

}


if(needsPrefix){
resolvedStyle=prefixAll(resolvedStyle);
}
return resolvedStyle;
};

module.exports=transformToWebStyle;