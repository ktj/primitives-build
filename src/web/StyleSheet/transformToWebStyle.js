const prefixAll = require('inline-style-prefixer/static'); // 2.4kb gzipped
const prefixedProperties = require('./prefixedProperties');
const normalizeValue = require('./normalizeValue');
const colorWithOpacity = require('../util/colorWithOpacity');
const hasOwnProperty = Object.prototype.hasOwnProperty;

const styleShortHands = {
  borderColor: {
    borderTopColor: true,
    borderRightColor: true,
    borderBottomColor: true,
    borderLeftColor: true,
  },
  borderRadius: {
    borderTopLeftRadius: true,
    borderTopRightRadius: true,
    borderBottomRightRadius: true,
    borderBottomLeftRadius: true,
  },
  borderStyle: {
    borderTopStyle: true,
    borderRightStyle: true,
    borderBottomStyle: true,
    borderLeftStyle: true,
  },
  borderWidth: {
    borderTopWidth: true,
    borderRightWidth: true,
    borderBottomWidth: true,
    borderLeftWidth: true,
  },
  margin: {
    marginTop: true,
    marginRight: true,
    marginBottom: true,
    marginLeft: true,
  },
  marginHorizontal: {
    marginRight: true,
    marginLeft: true,
  },
  marginVertical: {
    marginTop: true,
    marginBottom: true,
  },
  overflow: {
    overflowX: true,
    overflowY: true,
  },
  padding: {
    paddingTop: true,
    paddingRight: true,
    paddingBottom: true,
    paddingLeft: true,
  },
  paddingHorizontal: {
    paddingRight: true,
    paddingLeft: true,
  },
  paddingVertical: {
    paddingTop: true,
    paddingBottom: true,
  },
  textDecorationLine: {
    textDecoration: true,
  },
  writingDirection: {
    direction: true,
  },
};

/**
 * Alpha-sort properties, apart from shorthands â€“ they must appear before the
 * longhand properties that they expand into. This lets more specific styles
 * override less specific styles, whatever the order in which they were
 * originally declared.
 */
const sortProps = (propsArray) => propsArray.sort((a, b) => {
  const expandedA = styleShortHands[a];
  const expandedB = styleShortHands[b];
  if (expandedA && expandedA[b]) {
    return -1;
  } else if (expandedB && expandedB[a]) {
    return 1;
  }
  // eslint-disable-next-line no-nested-ternary
  return (a < b) ? -1 : (a > b) ? 1 : 0;
});

const resolveFlexStyle = (resolvedStyle, flex /* , style */) => {
  // TODO(lmr): RN now supports some of these properties. look into how to handle
  // eslint-disable-next-line no-param-reassign
  resolvedStyle.flexGrow = flex;
  // eslint-disable-next-line no-param-reassign
  resolvedStyle.flexShrink = 1;
  // eslint-disable-next-line no-param-reassign
  resolvedStyle.flexBasis = 'auto';
};

const resolveShadowProperty = (resolvedStyle, style) => {
  const opacity = style.shadowOpacity != null ? style.shadowOpacity : 1;
  const color = colorWithOpacity(style.shadowColor, opacity);
  const offset = style.shadowOffset || { width: 0, height: 0 };
  const x = normalizeValue('width', offset.width || 0);
  const y = normalizeValue('height', offset.height || 0);
  const r = style.shadowRadius || 0;

  // we multiply by 2 here because on native, the radius is in terms of
  // "points" which are multiple pixels, but on the web they are in terms of
  // pixels. A more same strategy might be to divide this by PizelRatio.get()
  // in the react native case.
  const rpx = normalizeValue('shadowRadius', r * 3);

  // eslint-disable-next-line no-param-reassign
  resolvedStyle.boxShadow = `${color} ${x} ${y} ${rpx}`;
};

// { scale: 2 } => 'scale(2)'
const mapTransform = (transform) => {
  const key = Object.keys(transform)[0];
  return `${key}(${normalizeValue(key, transform[key])})`;
};

// mutative
const resolveTransformProperty = (resolvedStyle, transform) => {
  // eslint-disable-next-line no-param-reassign
  resolvedStyle.transform = transform.map(mapTransform).join(' ');
};

/**
 * Expand the shorthand properties to isolate every declaration from the others.
 */
const transformToWebStyle = (style) => {
  if (!style) return style;

  const propsArray = Object.keys(style);
  const sortedProps = sortProps(propsArray);
  let resolvedStyle = {};
  let hasResolvedShadow = false;
  let needsPrefix = false;

  for (let i = 0; i < sortedProps.length; i++) {
    const key = sortedProps[i];
    const value = style[key];
    if (!needsPrefix && prefixedProperties[key]) {
      needsPrefix = true;
    }
    switch (key) {
      case 'flex':
        resolveFlexStyle(resolvedStyle, value, style);
        break;
      case 'textAlignVertical':
        resolvedStyle.verticalAlign = (value === 'center' ? 'middle' : value);
        break;
      case 'transform':
        resolveTransformProperty(resolvedStyle, value);
        break;
      case 'shadowColor':
      case 'shadowOffset':
      case 'shadowOpacity':
      case 'shadowRadius':
        if (!hasResolvedShadow) {
          resolveShadowProperty(resolvedStyle, style);
          hasResolvedShadow = true;
        }
        break;
      case 'borderColor':
      case 'borderRadius':
      case 'borderStyle':
      case 'borderWidth':
      case 'margin':
      case 'marginHorizontal':
      case 'marginVertical':
      case 'overflow':
      case 'padding':
      case 'paddingHorizontal':
      case 'paddingVertical':
      case 'textDecorationLine':
      case 'writingDirection': {
        const expandedProps = styleShortHands[key];
        const normalizedVal = normalizeValue(key, value);
        for (const propName in expandedProps) {
          if (hasOwnProperty.call(expandedProps, propName)) {
            resolvedStyle[propName] = normalizedVal;
          }
        }
      } break;
      case 'elevation':
        // ignore
        break;
      default:
        resolvedStyle[key] = normalizeValue(key, value);
        break;
    }
  }
  // micro-optimization. For simple styles that don't have properties that need to be prefixed,
  // we avoid making this call all together.
  if (needsPrefix) {
    resolvedStyle = prefixAll(resolvedStyle);
  }
  return resolvedStyle;
};

module.exports = transformToWebStyle;
