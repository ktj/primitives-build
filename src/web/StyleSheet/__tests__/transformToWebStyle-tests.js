import transformToWebStyle from '../transformToWebStyle';

describe('transformToWebStyle', () => {
  // TODO: boxShadow
  // TODO: transform
  // TODO: flex

  it('standard props', () => {
    expect(transformToWebStyle({
      width: 20,
      height: 20,
    })).toMatchSnapshot();
  });

  it('expanded props', () => {
    expect(transformToWebStyle({
      padding: 20,
      margin: 20,
    })).toMatchSnapshot();
  });

  it('flex 1', () => {
    expect(transformToWebStyle({
      flex: 1,
    })).toMatchSnapshot();
  });

  it('flex 0', () => {
    expect(transformToWebStyle({
      flex: 0,
    })).toMatchSnapshot();
  });

  it('flexbox properties', () => {
    expect(transformToWebStyle({
      flexDirection: 'row',
      justifyContent: 'center',
      alignItems: 'center',
    })).toMatchSnapshot();
  });

  describe('transform', () => {
    it('translateX', () => {
      expect(transformToWebStyle({
        transform: [
          { translateX: 20 },
        ],
      })).toMatchSnapshot();
    });

    it('rotateX', () => {
      expect(transformToWebStyle({
        transform: [
          { rotateX: '30deg' },
        ],
      })).toMatchSnapshot();
    });

    it('multiple', () => {
      expect(transformToWebStyle({
        transform: [
          { translateX: 20 },
          { rotateX: '30deg' },
        ],
      })).toMatchSnapshot();
    });
  });

  describe('expandable props', () => {
    it('paddingHorizontal', () => {
      expect(transformToWebStyle({
        paddingHorizontal: 10,
      })).toMatchSnapshot();
    });
    it('paddingVertical', () => {
      expect(transformToWebStyle({
        paddingVertical: 10,
      })).toMatchSnapshot();
    });
    it('padding', () => {
      expect(transformToWebStyle({
        padding: 10,
      })).toMatchSnapshot();
    });
    it('marginHorizontal', () => {
      expect(transformToWebStyle({
        marginHorizontal: 10,
      })).toMatchSnapshot();
    });
    it('marginVertical', () => {
      expect(transformToWebStyle({
        marginVertical: 10,
      })).toMatchSnapshot();
    });
    it('margin', () => {
      expect(transformToWebStyle({
        margin: 10,
      })).toMatchSnapshot();
    });
    it('borderWidth', () => {
      expect(transformToWebStyle({
        borderWidth: 10,
      })).toMatchSnapshot();
    });
    it('borderStyle', () => {
      expect(transformToWebStyle({
        borderStyle: 'solid',
      })).toMatchSnapshot();
    });
    it('borderColor', () => {
      expect(transformToWebStyle({
        borderColor: '#333',
      })).toMatchSnapshot();
    });
    it('borderRadius', () => {
      expect(transformToWebStyle({
        borderRadius: 10,
      })).toMatchSnapshot();
    });
  });

  describe('colors', () => {
    it('named', () => {
      expect(transformToWebStyle({
        color: 'blue',
      })).toMatchSnapshot();
    });

    it('rgb(...)', () => {
      expect(transformToWebStyle({
        color: 'rgb(10, 10, 10)',
      })).toMatchSnapshot();
    });

    it('rgba(...)', () => {
      expect(transformToWebStyle({
        color: 'rgba(10, 10, 10, 0.5)',
      })).toMatchSnapshot();
    });

    it('short hex', () => {
      expect(transformToWebStyle({
        color: '#fff',
      })).toMatchSnapshot();
    });

    it('normal hex', () => {
      expect(transformToWebStyle({
        color: '#fefefe',
      })).toMatchSnapshot();
    });

    it('long hex', () => {
      expect(transformToWebStyle({
        color: 'fefefefe',
      })).toMatchSnapshot();
    });
  });

  describe('boxShadow', () => {
    it('shadow 1', () => {
      expect(transformToWebStyle({
        shadowColor: '#000',
        shadowOpacity: 0.5,
        shadowRadius: 5,
        shadowOffset: {
          width: 1,
          height: 1,
        },
      })).toMatchSnapshot();
    });

    it('shadow 2', () => {
      expect(transformToWebStyle({
        shadowColor: '#000',
        shadowRadius: 5,
        shadowOffset: {
          width: 1,
          height: 1,
        },
      })).toMatchSnapshot();
    });

    it('shadow 3', () => {
      expect(transformToWebStyle({
        shadowColor: '#000',
        shadowOpacity: 0.5,
        shadowRadius: 5,
      })).toMatchSnapshot();
    });
  });

});
