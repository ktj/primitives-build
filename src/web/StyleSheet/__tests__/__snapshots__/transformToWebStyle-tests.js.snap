exports[`transformToWebStyle boxShadow shadow 1 1`] = `
Object {
  "boxShadow": "rgba(0,0,0,0.5) 1px 1px 15px",
}
`;

exports[`transformToWebStyle boxShadow shadow 2 1`] = `
Object {
  "boxShadow": "rgba(0,0,0,1) 1px 1px 15px",
}
`;

exports[`transformToWebStyle boxShadow shadow 3 1`] = `
Object {
  "boxShadow": "rgba(0,0,0,0.5) 0px 0px 15px",
}
`;

exports[`transformToWebStyle colors long hex 1`] = `
Object {
  "color": "fefefefe",
}
`;

exports[`transformToWebStyle colors named 1`] = `
Object {
  "color": "blue",
}
`;

exports[`transformToWebStyle colors normal hex 1`] = `
Object {
  "color": "#fefefe",
}
`;

exports[`transformToWebStyle colors rgb(...) 1`] = `
Object {
  "color": "rgb(10, 10, 10)",
}
`;

exports[`transformToWebStyle colors rgba(...) 1`] = `
Object {
  "color": "rgba(10, 10, 10, 0.5)",
}
`;

exports[`transformToWebStyle colors short hex 1`] = `
Object {
  "color": "#fff",
}
`;

exports[`transformToWebStyle expandable props borderColor 1`] = `
Object {
  "borderBottomColor": "#333",
  "borderLeftColor": "#333",
  "borderRightColor": "#333",
  "borderTopColor": "#333",
}
`;

exports[`transformToWebStyle expandable props borderRadius 1`] = `
Object {
  "borderBottomLeftRadius": "10px",
  "borderBottomRightRadius": "10px",
  "borderTopLeftRadius": "10px",
  "borderTopRightRadius": "10px",
}
`;

exports[`transformToWebStyle expandable props borderStyle 1`] = `
Object {
  "borderBottomStyle": "solid",
  "borderLeftStyle": "solid",
  "borderRightStyle": "solid",
  "borderTopStyle": "solid",
}
`;

exports[`transformToWebStyle expandable props borderWidth 1`] = `
Object {
  "borderBottomWidth": "10px",
  "borderLeftWidth": "10px",
  "borderRightWidth": "10px",
  "borderTopWidth": "10px",
}
`;

exports[`transformToWebStyle expandable props margin 1`] = `
Object {
  "marginBottom": "10px",
  "marginLeft": "10px",
  "marginRight": "10px",
  "marginTop": "10px",
}
`;

exports[`transformToWebStyle expandable props marginHorizontal 1`] = `
Object {
  "marginLeft": "10px",
  "marginRight": "10px",
}
`;

exports[`transformToWebStyle expandable props marginVertical 1`] = `
Object {
  "marginBottom": "10px",
  "marginTop": "10px",
}
`;

exports[`transformToWebStyle expandable props padding 1`] = `
Object {
  "paddingBottom": "10px",
  "paddingLeft": "10px",
  "paddingRight": "10px",
  "paddingTop": "10px",
}
`;

exports[`transformToWebStyle expandable props paddingHorizontal 1`] = `
Object {
  "paddingLeft": "10px",
  "paddingRight": "10px",
}
`;

exports[`transformToWebStyle expandable props paddingVertical 1`] = `
Object {
  "paddingBottom": "10px",
  "paddingTop": "10px",
}
`;

exports[`transformToWebStyle expanded props 1`] = `
Object {
  "marginBottom": "20px",
  "marginLeft": "20px",
  "marginRight": "20px",
  "marginTop": "20px",
  "paddingBottom": "20px",
  "paddingLeft": "20px",
  "paddingRight": "20px",
  "paddingTop": "20px",
}
`;

exports[`transformToWebStyle flex 0 1`] = `
Object {
  "WebkitFlexBasis": "auto",
  "WebkitFlexGrow": 0,
  "WebkitFlexShrink": 1,
  "flexBasis": "auto",
  "flexGrow": 0,
  "flexShrink": 1,
  "msFlexNegative": 1,
  "msFlexPositive": 0,
  "msPreferredSize": "auto",
}
`;

exports[`transformToWebStyle flex 1 1`] = `
Object {
  "WebkitFlexBasis": "auto",
  "WebkitFlexGrow": 1,
  "WebkitFlexShrink": 1,
  "flexBasis": "auto",
  "flexGrow": 1,
  "flexShrink": 1,
  "msFlexNegative": 1,
  "msFlexPositive": 1,
  "msPreferredSize": "auto",
}
`;

exports[`transformToWebStyle flexbox properties 1`] = `
Object {
  "WebkitAlignItems": "center",
  "WebkitBoxAlign": "center",
  "WebkitBoxDirection": "normal",
  "WebkitBoxOrient": "horizontal",
  "WebkitBoxPack": "center",
  "WebkitFlexDirection": "row",
  "WebkitJustifyContent": "center",
  "alignItems": "center",
  "flexDirection": "row",
  "justifyContent": "center",
  "msFlexAlign": "center",
  "msFlexDirection": "row",
  "msFlexPack": "center",
}
`;

exports[`transformToWebStyle standard props 1`] = `
Object {
  "height": "20px",
  "width": "20px",
}
`;

exports[`transformToWebStyle transform multiple 1`] = `
Object {
  "WebkitTransform": "translateX(20px) rotateX(30deg)",
  "msTransform": "translateX(20px) rotateX(30deg)",
  "transform": "translateX(20px) rotateX(30deg)",
}
`;

exports[`transformToWebStyle transform rotateX 1`] = `
Object {
  "WebkitTransform": "rotateX(30deg)",
  "msTransform": "rotateX(30deg)",
  "transform": "rotateX(30deg)",
}
`;

exports[`transformToWebStyle transform translateX 1`] = `
Object {
  "WebkitTransform": "translateX(20px)",
  "msTransform": "translateX(20px)",
  "transform": "translateX(20px)",
}
`;
